internal/app/singularity/verify_ocsp.go

```diff
@@ -0,0 +1,237 @@
+const (
+	PKIXOCSPNoCheck = "1.3.6.1.5.5.7.48.1.5"
+)
+
+var errOCSP = errors.New("OCSP verification has failed")
+
+func OCSPVerify(chain ...*x509.Certificate) error {
+	// use the pool as an index for certificate issuers.
+	// fixme: we can drop this lookup if we assume that certificate N is always signed by certificate N+1.
+	pool := map[string]*x509.Certificate{}
+
+	for _, cert := range chain {
+		pool[string(cert.SubjectKeyId)] = cert
+	}
+
+	// recursively validate the certificate chain
+	for _, cert := range chain {
+		if err := validateCertificate(cert, pool); err != nil {
+			sylog.Warningf("OCSP verification has failed. Err: %s", err)
+			return errOCSP
+		}
+	}
+
+	return nil
+}
+
+func validateCertificate(cert *x509.Certificate, pool map[string]*x509.Certificate) error {
+	if len(cert.AuthorityKeyId) == 0 || string(cert.SubjectKeyId) == string(cert.AuthorityKeyId) {
+		sylog.Infof("skip self-signed certificate (%s)", cert.Subject.String())
+
+		return nil
+	}
+
+	/*---------------------------------------------------
+	 * Retrieve the CA who issued the certificate in question.
+	 *---------------------------------------------------*/
+
+	// firstly, look for the issuer in the pool of certificates.
+	issuer, exists := pool[string(cert.AuthorityKeyId)]
+	if !exists {
+		// if not found anywhere locally, try to download it
+		missingCerts, err := downloadCertsFromURLs(cert.IssuingCertificateURL...)
+		if err != nil {
+			return errors.Wrapf(err, "download cert error")
+		}
+
+		// if that does not work either, just abort
+		issuer, exists = missingCerts[string(cert.AuthorityKeyId)]
+		if !exists {
+			return errors.Errorf("cannot find issuer '%s'", cert.Issuer)
+		}
+	}
+
+	sylog.Infof("Validate: cert:%s  issuer:%s", cert.Subject.CommonName, issuer.Subject.CommonName)
+
+	/*---------------------------------------------------
+	 * Ask OCSP for the validity of signer's certificate.
+	 * Also make sure that the OCSP is trustworthy.
+	 *---------------------------------------------------*/
+	ocspCertificate, err := queryOCSP(cert, issuer)
+	if err != nil {
+		return errors.Wrapf(err, "OCSP Query")
+	}
+
+	if ocspCertificate != nil {
+		// The CA requires us to explicitly trust this certificate
+		// RFC-6960 Section: 4.2.2.2.1
+		for _, extension := range cert.Extensions {
+			if extension.Id.String() == PKIXOCSPNoCheck {
+				goto skipOCSPVerification
+			}
+		}
+
+		// avoid loops cause by misconfigured OCSP responders
+		if string(cert.SubjectKeyId) == string(ocspCertificate.SubjectKeyId) {
+			return nil
+		}
+
+		// make sure that the OCSP server is trustworthy
+		if err := validateCertificate(ocspCertificate, pool); err != nil {
+			return errors.Wrapf(err, "cannot verify OCSP server's certificate")
+		}
+
+	skipOCSPVerification:
+	}
+
+	return nil
+}
+
+func queryOCSP(cert, issuer *x509.Certificate) (needsValidation *x509.Certificate, err error) {
+	logrus.Warnf("cert:[%s] issuer:[%s]", cert.Subject.String(), issuer.Subject.String())
```

Comment:
Should use `sylog` here:

```suggestion
	sylog.Debugf("cert:[%s] issuer:[%s]", cert.Subject.String(), issuer.Subject.String())
```